### PR TITLE
[Merged by Bors] - feat: port Tactic.Group

### PR DESCRIPTION
**Pull request by tb65536**
We implement `group` by alternating between a `simp` call and `ring_nf` (to normalize the exponents) until no progress is made.

To do this we also implement the tactic `fail_if_no_progress`, so that `repeat (fail_if_no_progress (simp ... <;> ring_nf))` behaves as expected.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Group.lean
Comment (mcdoll) [inline Mathlib/Tactic/Group.lean:101]: It is probably the same as
```suggestion
  `(tactic| (aux_group₁ $[at $location]?;
            aux_group₂ $[at $location]?;
            aux_group₁ $[at $location]?))
```
since you don't create additional goals.

Comment (tb65536) [reply Mathlib/Tactic/Group.lean:101]: It seems like the second and third tactics will sometimes complain that there are no goals to be solved.

**Review by j-loreaux — APPROVED** [review]
This looks good to me, and I checked locally that it works with some `group` invocations in #2143. Maybe it can be implemented more efficiently (e.g., because of the defeq checks in `fail_if_no_progress`), but I don't think we need to be particularly concerned with that right now. It seems to be fast enough.

**Comment by thorimur** [comment]
> This looks good to me, and I checked locally that it works with some `group` invocations in #2143. Maybe it can be implemented more efficiently (e.g., because of the defeq checks in `fail_if_no_progress`), but I don't think we need to be particularly concerned with that right now. It seems to be fast enough.

I agree that we shouldn't worry about it before merging this tactic; I'll continue enhancing fail_if_no_progress on a separate branch, but, since it's alright for now, I don't think we should slow down the port and hold up `group` by perfecting it prior to merging.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**
